### ISSUE TITLE
Learning behaviour when using pretrained model (n2v 3D)

### ISSUE THREAD

**Issue by nicesetup** [HenriquesLab/ZeroCostDL4Mic#50]
Hello @all,

first of all, thank you very much for providing this tool!
I am currently trying to use the n2v 3D notebook to denoise fluorescence images of calcium signals in order to make them useable for quantitative analysis.
For training, I use a stack of dimensions 512x512x250, with data augmentation enabled. Usually, the model converges somewhere between 50-150 epochs. However, I found that further training might still improve the results, even though the validation loss can not be further decreased by a significant amount. To examine how well validation loss correlates with the actual quality of predictions, I have adapted the code in the notebook to include a simple loop that after N epochs of training creates a prediction, then again loads the model and continues training for another N iterations and so on. This way I get to see "intermediate results" to better understand the training process. 
Pseudo-Code:
>create training and validation dataset
>do (num_of_epochs / N) times:
>....load pretrained model
>....train N epochs
>....save model and evaluation file
>....create prediction using current trained model
>end

However, when using a pretrained model, the loss function behaves in an unexpected way: Even though I load weights_last, validation loss changes drastically from the last epoch prior to model saving to the first epoch after re-loading the model. I have attached the log of a 200 epoch training, with intermediate results being generated every 25 epochs: [training_evaluation.pdf](https://github.com/HenriquesLab/ZeroCostDL4Mic/files/5333747/training_evaluation.pdf) Notice for example the aprupt increase of validation loss from epoch 75 to 76. Any ideas why this is happening? I have noticed the same behaviour when loading a pretrained model, e.g. to continue training after the 12 hour timeout of colab, so it does not seem to be related to my custom loop. If, however, training runs straight for 200 epochs without saving and reloading the model in between, the validation loss function behaves "normal".

Any help in understanding and/or improving this behaviour would be greatly appreciated!

Thank you in advance!

(the log file is in .pdf since it is not possible to upload .csv to github. In case you need the .csv file, I have uploaded it here: https://drive.google.com/file/d/1kmPj6SZtw-miZOgduZlRKjfye-cbHTou/view?usp=sharing)

**Comment by Romain-Laine** [comment]
Hi nicesetup,
What you describe is not surprising to me because of how the learning rate is handled. For your start and stop approach to be strictly equivalent to the single 200 epoch stretch of training, it would need to use exactly the same learning rate throughout.

Two important things about learning rate here:
- learning rate determines how large of an effect an error in the image output will have on the update of the model weights: a large learning rate may destroy what was learned previously and make the training/validation loss worse than at the last epoch of your pre-trained model
- when you reload your model, you need to make sure you use the learning rate that was used at that specific epoch where you stop. See what we do in our transfer learning section to get an idea of the code for that. This learning rate needs to be used as initial learning rate for your new training session.
- learning rate schedule (how it changes over the epochs) can be complex and in some cases needs some information about the previous few epochs. Therefore, even with the same initial learning rate it might still be different than the full session

So, in fewer words, please make sure that you change correctly the learning rate at each new training session, but this may not fix your problem.

An alternative would be to see if the CSBDeep models have a possibility of saving every single model (at each epoch) without stopping the training, and run the predictions on all the models after the training has ended. I am not sure yet whether that's possible. 

Have a try and let me know how you get on!

Romain

**Comment by Romain-Laine** [comment]
Quick update:
- I have checked and the current N2V models do not allow to save every single checkpoint during training, so that's not an option for now. 
- The N2V model uses ReduceLROnPlateau_ callback to adjust the learning rate and therefore will need to know the history of the learning rates to be adjusted. Therefore even changing the initial learning rate will not make your staggered training strictly equivalent.

Failing that, the only other solutions I currently see is
- to disable the learning rate update by setting train_reduce_lr to "None", that's not ideal and does not reflect what a real training session would do
- to record the learning rate during your full session and set the train_reduce_lr as ReduceLROnPlateau_ 'dict' for each small training session corresponding to the learning rates used during the full training session and hope for the best

I hope this helps !

Romain

**Comment by nicesetup** [comment]
Hello Romain,

thank you very much for your timely and detailed response!

First, sorry for not explaining that in enough detail: I did indeed load the last learning rate together with the weights from the previous iteration. To be precise, I use the code you implemented for transfer learning in the loop. That means that in the beginning of each loop iteration, I create a new config object for the model which also gets passed the last learning rate. You can see that this works from the log I attached, the drastic changes in validation loss occur even though the correct learning rate is chosen. 
Even if it is not strictly equivalent, I thought the only diffrence to a "real" training session would be that I affect the _patience_ parameter of ReduceLROnPlateu. I left it at the default value of 10, meaning that the Learning rate should be halved every 10 epochs if there are no further significant changes in the loss function. The only relevant history should therefore be that of the last 10 epochs. If now the model is saved and then reloaded, from my understanding, I would just "reset" the patience, meaning that the model would potentially take some extra epochs to progress in a meaningful way, since the ReduceLROnPlateau would kick in some epochs later than it normally would (since there is no history of the last 10 epochs). It is however not clear to me how this would cause the validation loss to spike (or decrease, which can also happen!) that drastically: The weights are the same, as is the learning rate, i.e. there should be no way to "destroy" what the model learned previously. Am I missing something here? Is it possible that something goes wrong by loading the weights in general? As I mentioned, the same behaviour can be observed outside my adapted code when just starting training from a pretrained model (on the same data with the same parameters).

Regarding model checkpoints after every epoch: Of course this would also be a viable approach, I don't necessarily need predictions during training, it would be sufficient to save a model checkpoint every N epochs and evaluate their performance later on. However, this is something that would probably have to be implemented in the Keras training process and I am not really familiar with that. Do you think this is possible in principle?

Thank you again very much for your help and best regards!

**Comment by Romain-Laine** [comment]
Hi,
OK, so you've definitely set your loop correctly from the learning rate point of view.
I am also surprised that it would change so much then.

Saving all or subset of checkpoints along the way is definitely possible with keras callbacks but this would need to modify the N2V current code. You can do that by cloning their repo, make modifications and get the N2V code from your new repo instead. 
N2V repo: https://github.com/juglab/n2v

Loading the weights simply is a built-in keras method so I don't expect it to go wrong there. You could try and compare the predictions from the last trained checkpoint and those from the pre-loaded network (without re-training), these should give you the same output.

Alternatively, you could try a different approach:
- train for 5 epochs from scratch --> get some predictions
- train for 10 epochs from scratch --> get some predictions
- train for 15 epochs from scratch --> get some predictions
etc.
Assuming that all trainings will roughly go down the same path, this should give you an idea of how it evolves. A better version of this would be to do that 5 or 10 times for each number of epochs and check how the predictions vary across a same number of epoch.
I have been meaning to test that for some time without having found the time to do so.

I hope this helps!

Romain

**Comment by nicesetup** [comment]
> You could try and compare the predictions from the last trained checkpoint and those from the pre-loaded network (without re-training), these should give you the same output.

You mean e.g. compare the prediction after my last loop iteration with the prediction I obtain by closing the colab session, reloading the model and then make a prediction with the same model without further training?
I am not sure whether I understood that correctly, but I think it might not be representative since for prediction weights_best are loaded in any case, whereas weights_last are used to continue training on the same model.

> Saving all or subset of checkpoints along the way is definitely possible with keras callbacks but this would need to modify the N2V current code. You can do that by cloning their repo, make modifications and get the N2V code from your new repo instead.
N2V repo: https://github.com/juglab/n2v

I will research on how to do this and get back to you, thanks!

> Alternatively, you could try a different approach:
>   train for 5 epochs from scratch --> get some predictions
>   train for 10 epochs from scratch --> get some predictions
>   train for 15 epochs from scratch --> get some predictions
>   etc.

While this might give me a better idea, always training from scratch might impact/falsify my conclusions since as you mention, the results may vary across the same number of epochs. Sometimes the model will converge earlier, say after 50 epochs, sometimes after 120 epochs. Also, I am afraid it is hardly feasable with the limitations of the free colab service, especially since I believe there might still be improvements after 150+ epochs.

Anyway, thank you very much for your quick responses!

**Comment by Romain-Laine** [comment]
What I meant about comparing predictions is only about checking that the weights are loaded properly in the code. Purely for testing that. 

For checkpointing your model weights, you can use ModelCheckpoint Keras callback:
https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/ModelCheckpoint

I agree with the issues you mention while training from scratch every time. Our approach of transfer learning aims at alleviating the runtime duration limit (hence you trying these!), so I'd be interested to hear how things have worked out on your data and training sessions!

Cheers,

Romain

**Comment by nicesetup** [comment]
Hello Romain,

just to quickly keep you updated: I performed the following experiment:
I trained a model for 150 epochs on my training sample and analyzed the resulting prediction. Then, I started training again on the same training sample, with the exact same parameters for **one** epoch using the "Using weights from a pre-trained model as initial weights" section (loading weights_last). The resulting model, which I would expect to perform very similar to the original model since it has been trained further for only a single epoch, produces a significantly worse prediction than the original model.
I repeated this process also for 5 and 10 epochs of training, always using the originally trained model as initial weights, and repeated each process four times to validate the result: The prediction is always significantly degreaded by starting from a pre-trained network. Training loss remained relatively constant compared to the original model, while validation loss again changed abruptly when starting training from the pretrained network. In this case that makes at least some amount of sense to me, since the validation dataset is diffrent for each trial, as it would normally be the case when resuming training from a saved model after a 12 hour timeout in colab. (This is in contrast to what I did in my loop described above, where I did not generate new training and validation datasets for each loop iteration.)

I conclude that there must be something missing when resuming training from a pretrained model. Some relevant parameter(s) influencing the training process, other than the learning rate, seem to get lost. Any ideas what might be the cause? As you mentioned, this issue might really be problematic since it means that the 12 hour runtime-constraint of colab can not be properly alleviated by saving and reloading model weights. Non-continuous training apparently produces strictly worse results (at least on my data).

Thank you very much for your time!

**Comment by Romain-Laine** [comment]
Hi,
Thanks for looking into that. That's really interesting. 
I have had another look at the code we have in ZeroCost and cannot currently find any issues with the way the weights or the LR are loaded into the new model for transfer learning.
As you suggest there may be another set of parameters that also needs to be fixed for this to work perfectly, but I cannot think what these might be right now (apart from keeping exactly the same training and validation split, but this would mean a poor generalisation of the model).

What I can tell you however is that N2V tends to be "unstable" with the first ~50 epochs, as we could often see from the training curves. I am not entirely sure why that is the case but there may be something going on under the hood that allows the model to explore large ranges of parameters space and therefore would somewhat make transfer learning inefficient.
I see that you have filed an issue with the N2V team, I'd also be interested to hear what their thoughts are, but I will take some time to look into this issue with our data and investigate!

Thanks for your thorough work!

Romain

**Comment by nicesetup** [comment]
Hi Romain,

just to keep you updated: The "missing parameter" is likely the state of the optimizer. Saving and reloading weights only seems reasonable for transfer learning, but for resuming training from the exact point where it was left off, it is insufficient to create a new model of the same structure and load only weights and learning rate. By doing that, you create a new optimizer object which of course does not have the same inner state as the previous one, resulting in diffrent training outcome. One would therefore have to save the entire model, including the optimizer, to alleviate computation-time contraints like the one in colab. (Correct me if you think I am wrong!)
I have also updated the issue I filed with the n2v team, but they seem to be pretty busy. I hope a solution to this problem can be implented in the near future.

Best regards,
nicesetup

**Comment by Romain-Laine** [comment]
Hi nicesetup, thanks for the update! 
I think you're perfectly right. We could consider saving the entire model and not just the weights, that's true. It's just the most common way of doing it (because only few people plan to use it the way we would for transfer learning here). 
I'll keep that in the back of my mind for our next releases, see what would be possible.

Cheers,

Romain

**Comment by nicesetup** [comment]
Dear Romain,
depending on the data you train on, n2v requires rather long training, such that stopping and continuing training is essential due to limitations in computation time. I think that keeping this issue more in the foreground of your mind for your next release would be benefitial for many users - note that you _suggest_ in your notebook to use the transfer learning section to alleviate the computation time constraint: 

> CRITICAL NOTE: Google Colab has a time limit for processing (to prevent using GPU power for datamining). Training time must be less than 12 hours! If training takes longer than 12 hours, please decrease the number of epochs or number of patches. Another way circumvent this is to save the parameters of the model after training and start training again from this point.

However, in it's current state, we can not use n2v to it's full potential becuase training for more than 12 hours might lead to decreased performance due to the reset of the optimizer.

Best regards,
nicesetup

**Comment by Romain-Laine** [comment]
Hi nicesetup,
I appreciate your comment. Please note that transfer learning seems to work fine for all other notebooks.
It's time-consuming to handle so many notebooks at the same time, so please be patient. 

Romain
